Fix `full` on symbolic value.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108166

Fix: #108067

This PR adds checks for `sympy.Expr` when extracting the dtype from a value inside the
`full` lowering.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ysiraichi committed:
Fix `full` on symbolic value.

Fix: #108067

This PR adds checks for `sympy.Expr` when extracting the dtype from a value inside the
`full` lowering.

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -131,6 +131,15 @@ def decode_dtype(dtype: int):
+def value_to_dtype(value: Any) -> torch.dtype:
+    if isinstance(value, sympy.Expr):
+        if value.is_integer is True:
+            return torch.long
```
Would it make more sense to do this as a decomposition so we can differentiate `SymInt` and `SymFloat`? Something like:

```python
if dtype is None:
    dtype = type_to_dtype(type(fill_value))
    return torch.full(...)
return NotImplemented
```

lezcano replied:
I think it's fine as is as a targeted fix for the issue. Further improvements can be done later on.

ysiraichi commented on torch/_inductor/lowering.py:
```diff
@@ -131,6 +131,15 @@ def decode_dtype(dtype: int):
 
+def value_to_dtype(value: Any) -> torch.dtype:
+    if isinstance(value, sympy.Expr):
+        if value.is_integer is True:
```
I guess it's not necessary here. It would be if we were checking for false, though.

ysiraichi committed:
Update on "Fix `full` on symbolic value."


Fix: #108067

This PR adds checks for `sympy.Expr` when extracting the dtype from a value inside the
`full` lowering.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge

ysiraichi committed:
Rebased. on "Fix `full` on symbolic value."


Fix: #108067

This PR adds checks for `sympy.Expr` when extracting the dtype from a value inside the
`full` lowering.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Fix `full` on symbolic value."


Fix: #108067

This PR adds checks for `sympy.Expr` when extracting the dtype from a value inside the
`full` lowering.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge
